### DISCUSSION TITLE
TypeScript Error in Discriminated Union

### DISCUSSION THREAD

**Discussion by emon5122** [colinhacks/zod#3131] [Q&A] [answered]
### Code
```typescript
const paymentMethods = [
  "COD",
  "CREDITCARD",
  "DEBITCARD",
] as const;

const methodsEnum = z.enum(paymentMethods);

const PaymentValidator = z.discriminatedUnion(
  "methods",
  paymentMethods.map((pm) => {
    if (pm !== "COD") {
      return z.object({
        methods: z.literal(methodsEnum.enum[pm]),
        amount: z.number(),
        transaction_id: z.string(),
      });
    }
    return z.object({
      methods: z.literal(methodsEnum.enum.COD),
      amount: z.number(),
    });
  }),
);
```
### Issue Description
The provided TypeScript code for a discriminated union using Zod is producing the following error:
Given code is showing typeScript error 
```sh
Argument of type '(ZodObject<{ methods: ZodLiteral< "CREDITCARD" | "DEBITCARD">; amount: ZodNumber; transaction_id: ZodString; }, "strip", ZodTypeAny, { ...; }, { ...; }> | ZodObject<...>)[]' is not assignable to parameter of type '[ZodDiscriminatedUnionOption<"methods">, ...ZodDiscriminatedUnionOption<"methods">[]]'.
Source provides no match for the required element at position 0 in the target.
```
However, the code seems to work as intended during runtime. When using PaymentValidator.safeParse, the validation results are as expected. Despite this, TypeScript errors persist.

### Expected Behavior
The TypeScript code should not produce the mentioned error, given that the runtime behavior appears correct.
### Additional Information
* Zod Version: 3.22.4
* TypeScript Version: 5.3.3

Any insights or suggestions to address this TypeScript error would be greatly appreciated. Thank you!

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const paymentMethods = [
    'COD',
    'CREDITCARD',
    'DEBITCARD',
] as const

const methodsEnum = z.enum( paymentMethods )

const PaymentValidator = z.discriminatedUnion( 'methods', [
    z.object( {
        methods: methodsEnum.exclude( [ 'COD' ] ),
        amount: z.number(),
        transaction_id: z.string(),
    } ),
    z.object( {
        methods: methodsEnum.extract( [ 'COD' ] ),
        amount: z.number(),
    } ),
] )
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger
